feat: control btc provider state via remote feature flag

fabiobozzo opened the pull request:
## **Description**

Implements Bitcoin provider feature flag support using the `addBitcoinAccount` remote feature flag. This allows the New Networks team to control Bitcoin account availability at runtime without requiring restarts.

**What is the reason for the change?**
We need a way to disable the Bitcoin provider via feature flag with the same behavior as Basic Functionality, but with more granular control over Bitcoin specifically.

**What is the improvement/solution?**
- Creates Bitcoin provider externally wrapped with `AccountProviderWrapper` for runtime control
- Subscribes to `RemoteFeatureFlagController:stateChange` to toggle provider state based on `addBitcoinAccount` flag
- Uses `wallet.sync()` to properly hide Bitcoin accounts when provider is disabled (no zombie accounts)
- Follows the exact approach proposed by the core team

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/36676?quickstart=1)

## **Changelog**

CHANGELOG entry: Added Bitcoin provider feature flag support for runtime control via addBitcoinAccount flag

## **Related issues**

[MUL-1138](https://consensyssoftware.atlassian.net/browse/MUL-1138)

## **Manual testing steps**

1. Enable `addBitcoinAccount` feature flag via remote config
2. Verify Bitcoin accounts can be created and are visible in account list
3. Disable `addBitcoinAccount` feature flag via remote config  
4. Verify Bitcoin accounts are hidden from UI (no longer appear in account list/groups)
5. Re-enable flag and verify Bitcoin accounts reappear
6. Confirm no zombie accounts remain when flag is disabled

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- Bitcoin provider was always enabled regardless of feature flag state -->

### **After**

<!-- Bitcoin provider respects addBitcoinAccount feature flag with runtime control -->

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I've included tests if applicable
- [x] I've documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I've applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

[MUL-1138]: https://consensyssoftware.atlassian.net/browse/MUL-1138?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Enables runtime control of the Bitcoin provider via version-gated remote flags and migrates BTC/SOL flags to new schema, updating selectors and tests accordingly.
> 
> - **Multichain Account Service**:
>   - Add `RemoteFeatureFlagController:stateChange` to allowed events and messenger typing.
>   - Wrap `BtcAccountProvider` with `AccountProviderWrapper` and toggle via remote flag; align wallets on enable.
> - **Feature Flags**:
>   - Introduce `shared/lib/multichain-feature-flags.ts` with `isMultichainFeatureEnabled` (version-gated helper).
>   - Migrate flags to object schema: `bitcoinAccounts`/`solanaAccounts` `{ enabled, minimumVersion }`.
>   - Bump `@metamask/remote-feature-flag-controller` to `^1.9.0`.
> - **Selectors/UI**:
>   - Update selectors to use version-gated checks for BTC/SOL support.
>   - Harden non‑EVM network filtering against missing configs.
> - **Tests/Mocks/Stories**:
>   - Update e2e/unit tests, stories, and mocks to new flag schema and BTC provider toggling.
>   - Adjust Solana tests to use `solanaAccounts` object flag.
> - **Lockfile**: Update `yarn.lock` for new dependency version.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit cfd33753069e320fe336dd3db4a1d1b681e20e12. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

Commit:
feat: control btc provider state via remote feature flag

- Added `BtcAccountProvider` and `AccountProviderWrapper` to the multichain account service.
- Integrated remote feature flag handling for enabling/disabling the Bitcoin provider based on state changes.
- Updated the messenger to subscribe to `RemoteFeatureFlagController:stateChange` for dynamic control of the Bitcoin provider's state.
- Ensured initial state is set based on the feature flag for adding Bitcoin accounts.

Commit:
fix lint issues

Commit:
debug btc provider initialization

Commit:
Merge branch 'main' into feat/MUL-1138-ff-disable-btc-provider

Commit:
fix rebase conflict

Commit:
Merge branch 'main' into feat/MUL-1138-ff-disable-btc-provider

Commit:
address pr feedback

Commit:
refactor: update RemoteFeatureFlagController subscription to handle state as an argument

Commit:
wrap getIsAddBitcoinAccountEnabled inside bitcoin build fence

Commit:
Merge branch 'main' into feat/MUL-1138-ff-disable-btc-provider

Commit:
address pr remarks

- Refactor `getIsAddBitcoinAccountEnabled` to use `RemoteFeatureFlagControllerState` for improved type safety.
- Rename variable `remoteFeatureFlagsState` to `initialRemoteFeatureFlagsState` for clarity.
- Update subscription handler to cast state to `RemoteFeatureFlagControllerState` for consistency.
- Enhance code readability and maintainability by streamlining feature flag checks.

Commit:
fix import order

Commit:
Merge branch 'main' into feat/MUL-1138-ff-disable-btc-provider

Commit:
feat: add version-aware Bitcoin feature flag checking

- Implement isAddBitcoinFlagEnabled utility with enabled/minVersion support
- Add version gating (Extension: >=13.6.0, Mobile: >=7.58.0)
- Remove code duplication by centralizing logic in selectors
- Fix ESLint errors (object destructuring, curly braces)
- Default to enabled when flag undefined for safe rollout
- Support both boolean flags and complex version objects

Commit:
fix tests expectations

Commit:
Merge branch 'main' into feat/MUL-1138-ff-disable-btc-provider

Commit:
fix ci failures and cursorbot remarks

Commit:
refactor: enhance version comparison logic

Commit:
streamline version retrieval

Commit:
fix require syntax

Commit:
address @aganglada pr remarks

Commit:
revert alignment and removal of btc accounts

Commit:
remove tron dummy impl

Commit:
moved import inside ff guard

Commit:
remove fallback legacy flag

Commit:
address @aganglada pr remarks and cursorbot

Commit:
iterate on ci failures

Commit:
iterate on ci failures

Commit:
fix: resolve Bitcoin feature flag crashes and update to bitcoinAccounts

- Fix createSelector undefined error for getIsBitcoinSupportEnabled
- Update addBitcoinAccount → bitcoinAccounts throughout codebase
- Add shared feature flag utility with version gating
- Update test mocks to proper flag structure

Commit:
revert erroneous version change

Commit:
feat: implement bitcoinAccounts feature flag with shared utilities

- Add shared multichain-feature-flags utility for consistent version gating
- Update all feature flag references from addBitcoinAccount to bitcoinAccounts
- Fix fence line syntax issues in conditional compilation blocks
- Update test mocks to use new bitcoinAccounts flag structure
- Resolve linting errors and formatting violations
- Ensure dynamic version loading from package.json
- Maintain minimal scope focusing only on provider enable/disable

Commit:
reenable alignment

Commit:
fix ci failures

aganglada commented on ui/selectors/multichain/networks.test.ts:
```diff
@@ -32,10 +32,12 @@ import {
   getIsSolanaSupportEnabled: jest.fn(
-    (state) => state.metamask.remoteFeatureFlags.addSolanaAccount,
+    (state) =>
+      state.metamask.remoteFeatureFlags.solanaAccounts?.enabled === true,
```
this one still didn't change

Commit:
address feedback

Commit:
address feedback

ccharly commented on app/scripts/controller-init/multichain/multichain-account-service-init.ts:
```diff
@@ -29,13 +45,20 @@ export const MultichainAccountServiceInit: ControllerInitFunction<
+      btcProvider,
       ///: END:ONLY_INCLUDE_IF
-    ],
+    ].filter(Boolean),
```
Are we ok with this then? 😁

```suggestion
    ]
```

Commit:
address ccharly feedback

Commit:
revert: remove unrelated formatting changes from identity test files

As requested by @ccharly - these files had minor formatting changes
that were unrelated to the Bitcoin feature flag implementation.

Commit:
remove unrelated changes

ccharly commented on ui/selectors/selectors.js:
```diff
@@ -3152,26 +3157,64 @@ export function getIsWatchEthereumAccountEnabled(state) {
-  return Boolean(addBitcoinAccount);
+  ///: BEGIN:ONLY_INCLUDE_IF(bitcoin)
+  const { bitcoinAccounts } = getRemoteFeatureFlags(state);
+  return isBitcoinAccountsFlagEnabled(bitcoinAccounts);
```
IMO, `isBitcoinAccountsFlagEnabled` should not expect the `bitcoinAccounts` object directly.

Instead, we should pass-in the `getRemoteFeatureFlags(state)` object, and `isBitcoinAccountsFlagEnabled` would use `bitcoinAccounts` from it (inside `isBitcoinAccountsFlagEnabled` I mean).

Because right now, I could be passing the result of `solanaAccounts` to `isBitcoinAccountsFlagEnabled` and that's ok, and I think it defeats the purpose of having named functions like this.

WDYT?

Commit:
refactor: address all Charlie's feedback - simplify feature flags

- Remove all unnecessary wrapper functions (isBitcoinAccountsFlagEnabled, etc.)
- Use isMultichainFeatureEnabled directly for cleaner architecture
- Implement previousValueComparator pattern for state change detection
- Use correct controllerMessenger for RemoteFeatureFlagController subscription
- Standardize on minimumVersion naming convention only

Addresses @ccharly feedback: https://github.com/MetaMask/metamask-extension/pull/36676

fabiobozzo commented on app/scripts/controller-init/multichain/multichain-account-service-init.ts:
```diff
@@ -79,6 +96,50 @@ export const MultichainAccountServiceInit: ControllerInitFunction<
+      );
+
+      // Only react if the flag actually changed
+      if (bitcoinAccountsEnabled !== currentBitcoinEnabled) {
```
I tried previousValueComparator but it has type signature issues with controllerMessenger.subscribe apparently, it expects `(payload: unknown) => void` but previousValueComparator returns `(value: RemoteFeatureFlagControllerState) => boolean`. The manual comparison approach works correctly as far as I know, instead

ccharly commented on shared/lib/multichain-feature-flags.ts:
```diff
@@ -0,0 +1,73 @@
+    );
+    return '99.99.99'; // High version number to pass version checks
+  }
+}
```
I would remove all those fallbacks TBH

Commit:
refactor: align btc ff with established multichain patterns

- Use string | null type matching MultichainAccountsFeatureFlag
- Remove complex fallbacks, use same APP_VERSION pattern
- Implement linter-friendly fence line structure
- Direct usage of isMultichainFeatureEnabled, no wrapper functions

Commit:
revert erroneous version change

ccharly committed:
feat: control btc provider state via remote feature flag (#36919)

## **Description**

Some fixes for:
- https://github.com/MetaMask/metamask-extension/pull/36676

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/36919?quickstart=1)

## **Changelog**

CHANGELOG entry: null

## **Related issues**

N/A

## **Manual testing steps**

N/A

## **Screenshots/Recordings**

### **Before**

### **After**

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask
Extension Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> <sup>[Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) is
generating a summary for commit
ad97fccca0e282945aa64ae97a8ec33ddd62e8f3. Configure
[here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

Commit:
fix version consistency

Commit:
fix version consistency

Commit:
fix prettier

Commit:
Merge branch 'main' into feat/MUL-1138-ff-disable-btc-provider

Commit:
Update app/scripts/controller-init/multichain/multichain-account-service-init.ts

Co-authored-by: Charly Chevalier <charly.chevalier@consensys.net>

ccharly commented on test/e2e/tests/solana/web-socket-connection.spec.ts:
```diff
@@ -51,7 +52,8 @@ describe('Solana Web Socket', function (this: Suite) {
         manifestFlags: {
           remoteFeatureFlags: {
-            addSolanaAccount: true,
+            addSolanaAccount: true, // NOTE: Will be depracated soon.
```
```suggestion
            addSolanaAccount: true, // NOTE: Will be deprecated soon.
```

ccharly commented on test/e2e/tests/solana/common-solana.ts:
```diff
@@ -1622,7 +1622,8 @@ export async function withSolanaAccountSnap(
         // - If this flag is set, the slides count will be 5.
         remoteFeatureFlags: {
-          addSolanaAccount: true,
+          addSolanaAccount: true, // NOTE: Will be depracated soon.
```
```suggestion
          addSolanaAccount: true, // NOTE: Will be deprecated soon.
```

Commit:
Apply suggestions from code review

Co-authored-by: Charly Chevalier <charly.chevalier@consensys.net>

Commit:
nitpic

ccharly committed:
feat: control btc provider state via remote feature flag (#36924)

## **Description**

More fixes/refactors.

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/36924?quickstart=1)

## **Changelog**

CHANGELOG entry: null

## **Related issues**

N/A

## **Manual testing steps**

N/A

## **Screenshots/Recordings**

### **Before**

### **After**

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask
Extension Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

ccharly commented on ui/selectors/assets.test.ts:
```diff
@@ -484,8 +484,10 @@ describe('getMultichainNativeAssetType', () => {
+        addSolanaAccount: true, // NOTE: Will soon be deprecated.
+        solanaAccounts: { enabled: true, minimumVersion: '13.6.0' },
+        addBitcoinAccount: true, // NOTE: Will soon be deprecated.
+        bitcoinAccounts: { enabled: true, minimumVersion: '13.6.0' },
```
Since we're removing the feature flag logic, let's cleanup those now actually:

```suggestion
        solanaAccounts: { enabled: true, minimumVersion: '13.6.0' },
        bitcoinAccounts: { enabled: true, minimumVersion: '13.6.0' },
```

ccharly commented on shared/lib/multichain-feature-flags.ts:
```diff
@@ -0,0 +1,58 @@
+ */
+export type MultichainFeatureFlag = {
+  enabled: boolean;
+  minimumVersion: string | null;
```
I think we can remove the `| null` actually, cause we really need it. WDYT?

Commit:
Update ui/selectors/assets.test.ts

Co-authored-by: Charly Chevalier <charly.chevalier@consensys.net>

Commit:
enforce a non-null minimumVersion

ccharly reviewed (approved):
LGTM! I left 2 suggestions for a follow-up:
- https://github.com/MetaMask/metamask-extension/pull/36676/files#r2436668544
- https://github.com/MetaMask/metamask-extension/pull/36676/files#r2436674128

But for now, the logic should work fine (assuming our feature flags are well-shaped obviously 🙃).
